### ISSUE TITLE
Shuffle may cause slow IO.

### ISSUE THREAD

**Issue by chinakook** [deepinsight/insightface#216]
The shuffling will cause slow IO because the file reader is out-of-order. This problem will occur even in the SSD disk.
https://github.com/deepinsight/insightface/blob/d37bb397f01256c83312675ca7a2cb3170b77ddc/src/image_iter.py#L93

I think remove shuffling or add a PrefetchingIter on the origin iter will speed up IO.
```
    data_train = FaceImageIter(...)
    data_val = FaceImageIter(...)
    data_train = mx.io.PrefetchingIter(data_train)
    data_val = mx.io.PrefetchingIter(data_val)
```

**Comment by YaqiLYU** [comment]
PrefetchingIter works well !
`    train_dataiter = FaceImageIter(
        batch_size           = args.batch_size,
        data_shape           = data_shape,
        path_imgrec          = path_imgrec,
        shuffle              = True,   #True
        rand_mirror          = args.rand_mirror,
        mean                 = mean,
        cutoff               = args.cutoff,
    )
    train_dataiter = mx.io.PrefetchingIter(train_dataiter)`

**Comment by chinakook** [comment]
@YaqiLYU has proved the PrefetchingIter take effect.

**Comment by YaqiLYU** [comment]
PrefetchingIter works! Add a line of code :
`    train_dataiter = FaceImageIter(
        batch_size           = args.batch_size,
        data_shape           = data_shape,
        path_imgrec          = path_imgrec,
        shuffle              = True,  
        rand_mirror          = args.rand_mirror,
        mean                 = mean,
        cutoff               = args.cutoff,
    )
    train_dataiter = mx.io.PrefetchingIter(train_dataiter)`

**Comment by kwanking** [comment]
用了确实效率有提升。

李大神建议数据io用dataloader, 见下面讨论
https://discuss.gluon.ai/t/topic/5235。 

不知道用这个效率是否会高些？

**Comment by ShiyangZhang** [comment]
with mx.io.PrefetchingIter it runs 20% faster than before. 
Removing random.shuffle(self.seq) has no effect for me.
Is there any way we can use RAM to pre-load FaceImageIter?

**Comment by nttstar** [comment]
code updated.

**Comment by chinakook** [comment]
Maybe.

**Comment by staceycy** [comment]
@chinakook Thank you for your quick response. 
In this way, maybe we can shuffle the data when generating rec file.

**Comment by jolastar** [comment]
My workaround for the slow IO performance with shuffling was to create a RAM disk and train on the dataset once copied to the RAM disk. 
Performance went up from ~50 images per second to ~500 images per second with a single Tesla T4 GPU. 
(And I got ~1,000 images per second with two T4 GPU's. I set the "per batch size" to 256 so with two GPU's I got the same batch size of 512 than the default configurations of 128 images per context with 4 GPU's.)
